controllers/operatorpolicy_controller.go

```diff
@@ -60,166 +65,205 @@ var _ reconcile.Reconciler = &OperatorPolicyReconciler{}
+	}
+
+	// Start query batch for caching and watching related objects
+	err := r.DynamicWatcher.StartQueryBatch(watcher)
+	if err != nil {
+		panic(err)
+	}
+
+	// Get the applied OperatorPolicy
+	err = r.Get(ctx, req.NamespacedName, policy)
 	if err != nil {
 		if errors.IsNotFound(err) {
 			OpLog.Info("Operator policy could not be found")
 
+			err := r.DynamicWatcher.EndQueryBatch(watcher)
+			if err != nil {
+				OpLog.Error(err, "Error ending the query batch in progress. Ignoring the failure.")
+			}
+
+			err = r.DynamicWatcher.RemoveWatcher(watcher)
+			if err != nil {
+				OpLog.Error(err, "Error updating dependency watcher. Ignoring the failure.")
+			}
+
 			return reconcile.Result{}, nil
 		}
+
+		OpLog.Error(err, "Failed to get operator policy")
+
+		return reconcile.Result{}, err
 	}
 
-	return reconcile.Result{}, nil
-}
+	// handle the policy
+	OpLog.Info("Reconciling OperatorPolicy", "policy", policy.Name)
 
-// PeriodicallyExecOperatorPolicies loops through all operatorpolicies in the target namespace and triggers
-// template handling for each one. This function drives all the work the operator policy controller does.
-func (r *OperatorPolicyReconciler) PeriodicallyExecOperatorPolicies(freq uint, elected <-chan struct{},
-) {
-	OpLog.Info("Waiting for leader election before periodically evaluating operator policies")
-	<-elected
+	// Check if specified Subscription exists
+	subscriptiongvk := schema.GroupVersionKind{Group: "operators.coreos.com", Version: "v1alpha1", Kind: "Subscription"}
 
-	exiting := false
-	for !exiting {
-		start := time.Now()
+	cachedItem, err := r.DynamicWatcher.Get(watcher, subscriptiongvk, policy.Spec.Subscription.Namespace,
+		policy.Spec.Subscription.SubscriptionSpec.Package)
+	if err != nil {
+		panic(err)
+	}
 
-		var skipLoop bool
-		if !skipLoop {
-			policiesList := policyv1beta1.OperatorPolicyList{}
+	subExists := cachedItem != nil
 
-			err := r.List(context.TODO(), &policiesList)
-			if err != nil {
-				OpLog.Error(err, "Failed to list the OperatorPolicy objects to evaluate")
-			} else {
-				for i := range policiesList.Items {
-					policy := policiesList.Items[i]
-					if !r.shouldEvaluatePolicy(&policy) {
-						continue
-					}
-
-					// handle policy
-					err := r.handleSinglePolicy(&policy)
-					if err != nil {
-						OpLog.Error(err, "Error while evaluating operator policy")
-					}
-				}
-			}
-		}
+	// Check if any OperatorGroups exist in the namespace
+	operatorgroupgvk := schema.GroupVersionKind{Group: "operators.coreos.com", Version: "v1", Kind: "OperatorGroup"}
 
-		elapsed := time.Since(start).Seconds()
-		if float64(freq) > elapsed {
-			remainingSleep := float64(freq) - elapsed
-			sleepTime := time.Duration(remainingSleep) * time.Second
-			OpLog.V(2).Info("Sleeping before reprocessing the operator policies", "seconds", sleepTime)
-			time.Sleep(sleepTime)
-		}
+	cachedList, err := r.DynamicWatcher.List(watcher, operatorgroupgvk, policy.Spec.OperatorGroup.Namespace, nil)
```

Comment:
If the OperatorGroup is nil, this could have a problem.

Reply:
Would changing this to the subscription namespace be fine? Since that the CSV is installed the in same namespace as the subscription, maybe we can use the subscription namespace to check for existing operatorGroups.